apps/glusterfs/operations_block_volume.go

```diff
@@ -376,6 +376,288 @@ func (bvc *BlockVolumeCreateOperation) CleanDone() error {
+		err := executor.BlockVolumeExpand(h, hvname, bv.Info.Name, bve.newSize)
+		if err != nil {
+			logger.LogError("Unable to Expand volume: %v", err)
+			return err
+		}
+		return nil
+	})
+}
+
+func (bve *BlockVolumeExpandOperation) Rollback(executor executors.Executor) error {
+	logger.Info("Starting Rollback for %v op:%v", bve.Label(), bve.op.Id)
+	return rollbackViaClean(bve, executor)
+}
+
+func (bve *BlockVolumeExpandOperation) Finalize() error {
+	return bve.db.Update(func(tx *bolt.Tx) error {
+		bv, e := NewBlockVolumeEntryFromId(tx, bve.bvolId)
+		if e != nil {
+			return e
+		}
+
+		bv.Info.Size = bve.newSize
+
+		bve.op.FinalizeBlockVolume(bv)
+		if e := bv.Save(tx); e != nil {
+			return e
+		}
+
+		bve.op.Delete(tx)
+		return nil
+	})
+}
+
+func (bve *BlockVolumeExpandOperation) Clean(executor executors.Executor) error {
+	logger.Info("Starting Clean for %v op:%v", bve.Label(), bve.op.Id)
+	var (
+		err error
+		bv  *BlockVolumeEntry
+		// hvname is the name of the hosting volume and is required
+		hvname string
+
+		// host mappings for exec-on-up-host try util
+		bvHosts nodeHosts
+	)
+	err = bve.db.View(func(tx *bolt.Tx) error {
+		txdb := wdb.WrapTx(tx)
+		bv, err = NewBlockVolumeEntryFromId(tx, bve.bvolId)
+		if err != nil {
+			return err
+		}
+		hvname, err = bv.blockHostingVolumeName(txdb)
+		if err != nil {
+			return err
+		}
+		bvHosts, err = bv.hosts(txdb)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+	if err != nil {
+		logger.LogError(
+			"failed to get state needed to get info of block volume: %v", err)
+		return err
+	}
+	// nothing past this point needs a db reference
+	logger.Info("executing get info of block volume %v in op:%v",
+		bve.bvolId, bve.op.Id)
+	err = newTryOnHosts(bvHosts).once().run(func(h string) error {
+		bvolInfo, err := bv.BlockVolumeInfoFromHost(executor, hvname, h)
+		if err != nil {
+			return err
+		}
+		bve.reclaim.actualSize = bvolInfo.Size
+		return nil
+	})
+
+	return err
+}
+
+func (bve *BlockVolumeExpandOperation) CleanDone() error {
+	logger.Info("Clean is done for %v op:%v", bve.Label(), bve.op.Id)
+	return bve.db.Update(func(tx *bolt.Tx) error {
+		bv, err := NewBlockVolumeEntryFromId(tx, bve.bvolId)
+		if err != nil {
+			return err
+		}
+		if bve.newSize != bve.reclaim.actualSize {
+			logger.Warning("Actual Size %v doesn't match newly requested Size %v",
+				bve.reclaim.actualSize, bve.newSize)
+			volume, err := NewVolumeEntryFromId(tx, bv.Info.BlockHostingVolume)
+			if err != nil {
+				return err
+			}
+			reclaimBhvSize := bve.newSize - bve.reclaim.actualSize
+			if err := volume.ModifyFreeSize(reclaimBhvSize); err != nil {
+				return err
+			}
+			err = volume.Save(tx)
+			if err != nil {
+				return err
+			}
+		} else {
+			logger.Info("Actual Size match newly requested Size %v", bve.reclaim.actualSize)
+			bv.Info.Size = bve.newSize
+			if err := bv.Save(tx); err != nil {
+				return err
+			}
+		}
+		if err := bve.op.Save(tx); err != nil {
```

Comment:
Should this be a delete of the operation instead of a save?

Reply:
I agree. This does look wrong to me now that you point it out. Let's fix this now.
This is another reason why I don't love doing the tests in a seperate PR because we should have a check in the tests that ensures no pending ops after the clean + clean done functions are run. Still not going to insist on that but this is why we do tests and not just reviews. :-)

Reply:
Good catch.